src/xrpld/app/tx/detail/AMMWithdraw.h

```diff
@@ -80,33 +80,6 @@ class AMMWithdraw : public Transactor
     TER
     doApply() override;
 
-private:
-    std::pair<TER, bool>
-    applyGuts(Sandbox& view);
-
-    /** Withdraw requested assets and token from AMM into LP account.
-     * Return new total LPToken balance.
-     * @param view
-     * @param ammAccount
-     * @param amountBalance
-     * @param amountWithdraw
-     * @param amount2Withdraw
-     * @param lpTokensAMMBalance current AMM LPT balance
-     * @param lpTokensWithdraw
-     * @param tfee
-     * @return
-     */
-    std::pair<TER, STAmount>
-    withdraw(
```

Comment:
can we keep withdrawAll as bool? Otherwise, it is kind of verbose to use
`WithdrawAll withdrawAll = ctx_.tx[sfFlags] & (tfWithdrawAll | tfOneAssetWithdrawAll) ? WithdrawAll::Yes : WithdrawAll::No;`
instead of `bool withdrawAll = ctx_.tx[sfFlags] & (tfWithdrawAll | tfOneAssetWithdrawAll)`

Reply:
It's just more clear when you call this function with enum as opposed to true/false. Perhaps you can add a helper function, which takes the flags and returns enum to make it less verbose?

Reply:
Also, in `withdraw` function, it checks the `lpTokens` hold by the `account`.
`auto const lpTokens = ammLPHolds(view, ammSle, account, journal);` 
If we pass `issuer` as `account`, then it gets the issuer's lptokens, which will affect the following logic which uses this `lpTokens` variable. For example, the issuer does not hold any lptokens, and we possibly will return as `lpTokensWithdrawActual > lpTokens`.

Reply:
Good point. What if the `withdrawAll` is extended: `enum WithdrawAll {Yes, TransferToIssuer, TransferAllToIssuer, No}`? Anything but `No` means withdrawAll. If `TransferToIssuer` is set then `amount` is transferred to the issuer and if `TransferAllToIssuer` is set then `amount2` is transferred to the issuer. The issuer account doesn't have to be included since it's included in the amounts.

Reply:
Yeah, we can make this change

Reply:
@gregtatcam
maybe we can rename `WithdrawAll` to `WithdrawOption`?
`enum WithdrawOption {Yes, TransferToIssuer, TransferAllToIssuer, No}`

I have another concern,
Since withdraw is used in many places, is it better to keep its function as simple as possible? And clawback change is on top of this layer by using its return value.
Of course, by adding logics directly into the withdraw function is another option and makes less code change in caller functions. But this will make `withdraw` more complicated. And make it check the `WithdrawOption` each time. What do you think?

Reply:
Renaming sounds good to me.

We either have a common function or have two separate functions. The latter is not an option. Since the common function has to work for both use-cases, it does become more complicated one way or another. To mitigate this complexity in AMMWithdraw, we are adding a private non-static `withdraw`, which is going to hide most of the changes from the callers. I'm ok with returning amounts, again given the overload, if you feel strongly about it. Though I would argue that transferring to the holder and then to the issuer is not ideal either.

Reply:
If we continue use `WithdrawOption`, we have to separate it from `WithdrawAll`. 
`enum WithdrawOption {BothToAccount, AmountToIssuer, BothToIssuer}` is to control if sending to issuer. `enum WithdrawAll {true, false}` is to control if withdraw all.
Because the following situations will be a 6 combinations of the usage:
1. withdraw all to account
2. not withdraw all to account
3. withdraw all + token A to issuer + token B to account
4. not withdraw all + token A to issuer + token B to account
5. withdraw all  + token A and B to issuer
6. not withdraw all + token A and B to issuer

This will make the function taking another parameter and more complex. And all the callers need to be very clear about which flow it's going to take.
Basically for withdraw function, I think it would be better if it just take care of the withdraw logics like: "withdraw to an account."
And returning the withdrawn amounts will be more extensible in the future if these amounts are needed again in another feature.

Reply:
Sounds good. But let's have a simplified private overload for other AMMWithdraw functions to call.